src/Microsoft.AspNetCore.Blazor/Rendering/Renderer.cs

```diff
@@ -77,7 +79,8 @@ private ComponentState AttachAndInitComponent(IComponent component, int parentCo
         /// Updates the visible UI.
         /// </summary>
         /// <param name="renderBatch">The changes to the UI since the previous call.</param>
-        protected abstract void UpdateDisplay(in RenderBatch renderBatch);
+        /// <returns>A <see cref="Task"/> to represent the UI update process.</returns>
+        protected abstract Task UpdateDisplay(in RenderBatch renderBatch);
```

Comment:
This is the central change in this PR. The renderer now has a way to know when the client later finishes processing the `UpdateDisplay` request. It's no longer just "fire and forget".